File: backend/scripts/finetune_openai.py
#!/usr/bin/env python3
"""
Загрузка датасета и запуск fine-tuning gpt-4o-mini через OpenAI API.

Датасет готовится отдельно (см. export-finetune-dataset.js) — этот скрипт
только загружает уже готовый .jsonl файл, создаёт задание на обучение и
следит за его статусом до завершения.

Установка зависимостей (один раз):
    pip install openai python-dotenv

Использование:
    python3 finetune_openai.py --file ../training-data/finetune-dataset.jsonl \
        --base-model gpt-4o-mini-2024-07-18 \
        --suffix vika-v1

По завершении скрипт печатает готовый model id вида:
    ft:gpt-4o-mini-2024-07-18:org::abc123

Этот id нужно вписать в OPENAI_MODEL в .env бота (см. docs/FINE_TUNING.md).
"""

import argparse
import json
import os
import sys
import time
from pathlib import Path

try:
    from dotenv import load_dotenv
except ImportError:
    load_dotenv = None

try:
    from openai import OpenAI
except ImportError:
    print("Пакет openai не установлен. Выполни: pip install openai python-dotenv")
    sys.exit(1)


def load_env():
    """Подгружает .env из backend/.env, если python-dotenv установлен."""
    if load_dotenv is None:
        print("[warn] python-dotenv не установлен — переменные окружения берутся только из shell.")
        return
    env_path = Path(__file__).resolve().parent.parent / ".env"
    if env_path.exists():
        load_dotenv(env_path)
        print(f"[finetune] Загружен .env из {env_path}")
    else:
        print(f"[warn] .env не найден по пути {env_path} — используются переменные из shell.")


def build_client() -> OpenAI:
    api_key = os.environ.get("OPENAI_API_KEY")
    if not api_key:
        print("Ошибка: OPENAI_API_KEY не задан ни в .env, ни в окружении.")
        sys.exit(1)
    base_url = os.environ.get("OPENAI_BASE_URL") or None
    # ВАЖНО: если бот работает через SOCKS5-прокси (см. aiResponder.js,
    # api.openai.com блокирует российские IP) — на машине, где запускается
    # ЭТОТ скрипт, тоже нужен доступ к api.openai.com. Проще всего гонять
    # fine-tuning с сервера/машины, у которой есть прямой доступ (например,
    # через VPN), либо экспортировать HTTPS_PROXY/ALL_PROXY перед запуском:
    #   export ALL_PROXY=socks5://user:pass@host:port
    #   python3 finetune_openai.py ...
    # httpx (используется openai-python) сам подхватывает эти переменные.
    kwargs = {"api_key": api_key}
    if base_url:
        kwargs["base_url"] = base_url
        print(f"[finetune] Базовый URL переопределён: {base_url}")
    return OpenAI(**kwargs)


def validate_jsonl(path: Path) -> int:
    """Быстрая локальная проверка формата перед отправкой — экономит время
    и деньги на случай кривого датасета."""
    count = 0
    with path.open("r", encoding="utf-8") as f:
        for i, line in enumerate(f, start=1):
            line = line.strip()
            if not line:
                continue
            try:
                obj = json.loads(line)
            except json.JSONDecodeError as e:
                print(f"[error] Строка {i}: невалидный JSON — {e}")
                sys.exit(1)
            messages = obj.get("messages")
            if not isinstance(messages, list) or len(messages) < 2:
                print(f"[error] Строка {i}: поле 'messages' должно быть списком минимум из 2 сообщений")
                sys.exit(1)
            roles = [m.get("role") for m in messages]
            if "assistant" not in roles:
                print(f"[error] Строка {i}: в примере нет сообщения с role='assistant'")
                sys.exit(1)
            count += 1
    if count < 10:
        print(f"[error] В датасете {count} примеров — OpenAI требует минимум 10 для запуска fine-tuning.")
        sys.exit(1)
    print(f"[finetune] Локальная проверка пройдена: {count} валидных примеров.")
    return count


def upload_file(client: OpenAI, path: Path) -> str:
    print(f"[finetune] Загружаю файл {path} ...")
    with path.open("rb") as f:
        uploaded = client.files.create(file=f, purpose="fine-tune")
    print(f"[finetune] Файл загружен, file_id = {uploaded.id}")
    return uploaded.id


def create_job(client: OpenAI, file_id: str, base_model: str, suffix: str | None) -> str:
    kwargs = {"training_file": file_id, "model": base_model}
    if suffix:
        kwargs["suffix"] = suffix
    job = client.fine_tuning.jobs.create(**kwargs)
    print(f"[finetune] Задание создано: job_id = {job.id}, статус = {job.status}")
    return job.id


def poll_job(client: OpenAI, job_id: str, poll_seconds: int = 30):
    """Опрашивает статус задания до финального состояния (succeeded/failed/cancelled)."""
    terminal_states = {"succeeded", "failed", "cancelled"}
    while True:
        job = client.fine_tuning.jobs.retrieve(job_id)
        print(f"[finetune] Статус: {job.status} (обновлено {time.strftime('%H:%M:%S')})")

        # Печатаем последние события — удобно видеть прогресс эпох обучения.
        try:
            events = client.fine_tuning.jobs.list_events(fine_tuning_job_id=job_id, limit=3)
            for ev in reversed(list(events.data)):
                print(f"    · {ev.message}")
        except Exception:
            pass

        if job.status in terminal_states:
            return job
        time.sleep(poll_seconds)


def main():
    parser = argparse.ArgumentParser(description="Запуск fine-tuning gpt-4o-mini на датасете диалогов бота.")
    parser.add_argument("--file", required=True, help="Путь к .jsonl датасету (см. export-finetune-dataset.js)")
    parser.add_argument(
        "--base-model",
        default="gpt-4o-mini-2024-07-18",
        help="Базовая модель для fine-tuning (по умолчанию gpt-4o-mini-2024-07-18)",
    )
    parser.add_argument("--suffix", default=None, help="Короткий суффикс для имени модели (напр. 'vika-v1')")
    parser.add_argument(
        "--poll-seconds", type=int, default=30, help="Интервал опроса статуса задания в секундах (по умолчанию 30)"
    )
    parser.add_argument(
        "--no-wait",
        action="store_true",
        help="Не ждать завершения — только создать задание и вывести job_id (проверять статус отдельно)",
    )
    args = parser.parse_args()

    load_env()

    dataset_path = Path(args.file).resolve()
    if not dataset_path.exists():
        print(f"Файл не найден: {dataset_path}")
        sys.exit(1)

    validate_jsonl(dataset_path)

    client = build_client()
    file_id = upload_file(client, dataset_path)
    job_id = create_job(client, file_id, args.base_model, args.suffix)

    if args.no_wait:
        print(f"\n[finetune] Задание запущено в фоне. Проверить статус позже:")
        print(f"    python3 finetune_openai.py --check {job_id}")
        print(f"job_id = {job_id}")
        return

    job = poll_job(client, job_id, args.poll_seconds)

    print("\n" + "=" * 60)
    if job.status == "succeeded":
        print(f"ГОТОВО! Fine-tuned модель обучена успешно.")
        print(f"model_id: {job.fine_tuned_model}")
        print(f"\nЧтобы использовать её в боте, впиши в backend/.env:")
        print(f"    OPENAI_MODEL={job.fine_tuned_model}")
        print(f"и перезапусти бота: pm2 restart all")
        if getattr(job, "trained_tokens", None):
            print(f"\nОбучено токенов: {job.trained_tokens}")
            print(
                f"Ориентировочная стоимость обучения (по цене ~$3.00 / 1M токенов для gpt-4o-mini): "
                f"${job.trained_tokens / 1_000_000 * 3.0:.2f} "
                f"(уточни актуальную цену на platform.openai.com/pricing)"
            )
    else:
        print(f"Задание завершилось со статусом: {job.status}")
        if getattr(job, "error", None):
            print(f"Ошибка: {job.error}")
    print("=" * 60)


if __name__ == "__main__":
    main()

File: backend/scripts/finetune-openai.js
#!/usr/bin/env node
/**
 * Загрузка датасета и запуск fine-tuning gpt-4o-mini через OpenAI API.
 *
 * Node.js-версия — использует тот же пакет `openai`, что и сам бот
 * (services/aiResponder.js), поэтому не требует Python/pip: весь пайплайн
 * fine-tuning остаётся на одном языке с проектом.
 *
 * Датасет готовится отдельно (см. export-finetune-dataset.js) — этот скрипт
 * только загружает уже готовый .jsonl файл, создаёт задание на обучение и
 * следит за его статусом до завершения.
 *
 * Использование:
 *   cd backend
 *   node scripts/finetune-openai.js \
 *     --file=../training-data/finetune-dataset.jsonl \
 *     --base-model=gpt-4o-mini-2024-07-18 \
 *     --suffix=vika-v1
 *
 * По завершении скрипт печатает готовый model id вида:
 *   ft:gpt-4o-mini-2024-07-18:org::abc123
 *
 * Этот id нужно вписать в OPENAI_MODEL в .env бота (см. docs/FINE_TUNING.md).
 *
 * Флаги:
 *   --file          путь к .jsonl датасету (обязателен)
 *   --base-model    базовая модель (по умолчанию gpt-4o-mini-2024-07-18)
 *   --suffix        короткий суффикс для имени модели (напр. "vika-v1")
 *   --poll-seconds  интервал опроса статуса задания (по умолчанию 30)
 *   --no-wait       не ждать завершения — только создать задание и вывести job_id
 *   --check=JOB_ID  не создавать новое задание — только проверить статус существующего
 */

require('dotenv').config();
const fs = require('fs');
const path = require('path');
const OpenAI = require('openai');

function parseArgs() {
  const args = {};
  for (const arg of process.argv.slice(2)) {
    const m = arg.match(/^--([^=]+)=?(.*)$/);
    if (m) args[m[1]] = m[2] === '' ? true : m[2];
  }
  return args;
}

function buildClient() {
  const apiKey = process.env.OPENAI_API_KEY;
  if (!apiKey) {
    console.error('Ошибка: OPENAI_API_KEY не задан ни в .env, ни в окружении.');
    process.exit(1);
  }
  const baseURL = process.env.OPENAI_BASE_URL || undefined;
  // ВАЖНО: если бот работает через SOCKS5-прокси (см. aiResponder.js,
  // api.openai.com блокирует российские IP) — на машине, где запускается
  // ЭТОТ скрипт, тоже нужен доступ к api.openai.com. Проще всего гонять
  // fine-tuning с сервера/машины с прямым доступом (например, через VPN),
  // либо задать HTTPS_PROXY/ALL_PROXY перед запуском — пакет `openai`
  // сам их не подхватывает, поэтому при необходимости используйте
  // `httpAgent` (см. aiResponder.js) через переменную OPENAI_PROXY_URL.
  const kwargs = { apiKey };
  if (baseURL) {
    kwargs.baseURL = baseURL;
    console.log(`[finetune] Базовый URL переопределён: ${baseURL}`);
  }
  if (process.env.OPENAI_PROXY_URL) {
    const { SocksProxyAgent } = require('socks-proxy-agent');
    kwargs.httpAgent = new SocksProxyAgent(process.env.OPENAI_PROXY_URL);
    console.log('[finetune] Использую SOCKS5-прокси из OPENAI_PROXY_URL.');
  }
  return new OpenAI(kwargs);
}

/** Быстрая локальная проверка формата перед отправкой — экономит время и
 * деньги на случай кривого датасета. */
function validateJsonl(filePath) {
  const content = fs.readFileSync(filePath, 'utf8');
  const lines = content.split('\n').filter((l) => l.trim() !== '');
  let count = 0;
  lines.forEach((line, idx) => {
    let obj;
    try {
      obj = JSON.parse(line);
    } catch (e) {
      console.error(`[error] Строка ${idx + 1}: невалидный JSON — ${e.message}`);
      process.exit(1);
    }
    const messages = obj.messages;
    if (!Array.isArray(messages) || messages.length < 2) {
      console.error(`[error] Строка ${idx + 1}: поле 'messages' должно быть списком минимум из 2 сообщений`);
      process.exit(1);
    }
    const roles = messages.map((m) => m.role);
    if (!roles.includes('assistant')) {
      console.error(`[error] Строка ${idx + 1}: в примере нет сообщения с role='assistant'`);
      process.exit(1);
    }
    count += 1;
  });
  if (count < 10) {
    console.error(`[error] В датасете ${count} примеров — OpenAI требует минимум 10 для запуска fine-tuning.`);
    process.exit(1);
  }
  console.log(`[finetune] Локальная проверка пройдена: ${count} валидных примеров.`);
  return count;
}

async function uploadFile(client, filePath) {
  console.log(`[finetune] Загружаю файл ${filePath} ...`);
  const uploaded = await client.files.create({
    file: fs.createReadStream(filePath),
    purpose: 'fine-tune',
  });
  console.log(`[finetune] Файл загружен, file_id = ${uploaded.id}`);
  return uploaded.id;
}

async function createJob(client, fileId, baseModel, suffix) {
  const payload = { training_file: fileId, model: baseModel };
  if (suffix) payload.suffix = suffix;
  const job = await client.fineTuning.jobs.create(payload);
  console.log(`[finetune] Задание создано: job_id = ${job.id}, статус = ${job.status}`);
  return job.id;
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/** Опрашивает статус задания до финального состояния (succeeded/failed/cancelled). */
async function pollJob(client, jobId, pollSeconds) {
  const terminalStates = new Set(['succeeded', 'failed', 'cancelled']);
  while (true) {
    const job = await client.fineTuning.jobs.retrieve(jobId);
    console.log(`[finetune] Статус: ${job.status} (обновлено ${new Date().toLocaleTimeString()})`);

    try {
      const events = await client.fineTuning.jobs.listEvents(jobId, { limit: 3 });
      [...events.data].reverse().forEach((ev) => console.log(`    · ${ev.message}`));
    } catch {
      // не критично — просто пропускаем вывод событий
    }

    if (terminalStates.has(job.status)) return job;
    await sleep(pollSeconds * 1000);
  }
}

async function main() {
  const args = parseArgs();

  const client = buildClient();

  // Режим проверки существующего задания без создания нового.
  if (args.check) {
    const job = await client.fineTuning.jobs.retrieve(args.check);
    console.log(`[finetune] Статус задания ${args.check}: ${job.status}`);
    if (job.status === 'succeeded') {
      console.log(`model_id: ${job.fine_tuned_model}`);
    } else if (job.error) {
      console.log(`Ошибка: ${JSON.stringify(job.error)}`);
    }
    return;
  }

  if (!args.file) {
    console.error('Ошибка: укажи --file=путь/к/датасету.jsonl');
    process.exit(1);
  }

  const datasetPath = path.resolve(process.cwd(), args.file);
  if (!fs.existsSync(datasetPath)) {
    console.error(`Файл не найден: ${datasetPath}`);
    process.exit(1);
  }

  validateJsonl(datasetPath);

  const baseModel = args['base-model'] || 'gpt-4o-mini-2024-07-18';
  const suffix = args.suffix || null;
  const pollSeconds = Number(args['poll-seconds'] ?? 30);

  const fileId = await uploadFile(client, datasetPath);
  const jobId = await createJob(client, fileId, baseModel, suffix);

  if (args['no-wait']) {
    console.log('\n[finetune] Задание запущено в фоне. Проверить статус позже:');
    console.log(`    node scripts/finetune-openai.js --check=${jobId}`);
    console.log(`job_id = ${jobId}`);
    return;
  }

  const job = await pollJob(client, jobId, pollSeconds);

  console.log('\n' + '='.repeat(60));
  if (job.status === 'succeeded') {
    console.log('ГОТОВО! Fine-tuned модель обучена успешно.');
    console.log(`model_id: ${job.fine_tuned_model}`);
    console.log('\nЧтобы использовать её в боте, впиши в backend/.env:');
    console.log(`    OPENAI_MODEL=${job.fine_tuned_model}`);
    console.log('и перезапусти бота: pm2 restart all');
    if (job.trained_tokens) {
      const approxCost = (job.trained_tokens / 1_000_000) * 3.0;
      console.log(`\nОбучено токенов: ${job.trained_tokens}`);
      console.log(
        `Ориентировочная стоимость обучения (по цене ~$3.00 / 1M токенов для gpt-4o-mini): $${approxCost.toFixed(2)} ` +
          '(уточни актуальную цену на platform.openai.com/pricing)',
      );
    }
  } else {
    console.log(`Задание завершилось со статусом: ${job.status}`);
    if (job.error) console.log(`Ошибка: ${JSON.stringify(job.error)}`);
  }
  console.log('='.repeat(60));
}

main().catch((err) => {
  console.error('[finetune] Ошибка:', err);
  process.exit(1);
});
